### PR TITLE
[SPARK-23207][SPARK-22905][SPARK-24564][SPARK-25114][SQL][BACKPORT-2.2] Shuffle+Repartition on a DataFrame could lead to incorrect answers

### PR DESCRIPTION
**Pull request by bersprockets**
## What changes were proposed in this pull request?

Back port of #20393.

Currently shuffle repartition uses RoundRobinPartitioning, the generated result is nondeterministic since the sequence of input rows are not determined.

The bug can be triggered when there is a repartition call following a shuffle (which would lead to non-deterministic row ordering), as the pattern shows below:
upstream stage -> repartition stage -> result stage
(-> indicate a shuffle)
When one of the executors process goes down, some tasks on the repartition stage will be retried and generate inconsistent ordering, and some tasks of the result stage will be retried generating different data.

The following code returns 931532, instead of 1000000:
```
import scala.sys.process._

import org.apache.spark.TaskContext
val res = spark.range(0, 1000 * 1000, 1).repartition(200).map { x =>
  x
}.repartition(200).map { x =>
  if (TaskContext.get.attemptNumber == 0 && TaskContext.get.partitionId < 2) {
    throw new Exception("pkill -f java".!!)
  }
  x
}
res.distinct().count()
```

In this PR, we propose a most straight-forward way to fix this problem by performing a local sort before partitioning, after we make the input row ordering deterministic, the function from rows to partitions is fully deterministic too.

The downside of the approach is that with extra local sort inserted, the performance of repartition() will go down, so we add a new config named `spark.sql.execution.sortBeforeRepartition` to control whether this patch is applied. The patch is default enabled to be safe-by-default, but user may choose to manually turn it off to avoid performance regression.

This patch also changes the output rows ordering of repartition(), that leads to a bunch of test cases failure because they are comparing the results directly.

Add unit test in ExchangeSuite.

With this patch(and `spark.sql.execution.sortBeforeRepartition` set to true), the following query returns 1000000:
```
import scala.sys.process._

import org.apache.spark.TaskContext

spark.conf.set("spark.sql.execution.sortBeforeRepartition", "true")

val res = spark.range(0, 1000 * 1000, 1).repartition(200).map { x =>
  x
}.repartition(200).map { x =>
  if (TaskContext.get.attemptNumber == 0 && TaskContext.get.partitionId < 2) {
    throw new Exception("pkill -f java".!!)
  }
  x
}
res.distinct().count()

res7: Long = 1000000
```

Author: Xingbo Jiang <xingbo.jiang@databricks.com>

## How was this patch tested?

Ran all SBT unit tests for org.apache.spark.sql.*.

Ran pyspark tests for module pyspark-sql.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @jiangxb1987

**Comment by jiangxb1987** [comment]
We shall also include #20088 in this backport PR.

**Comment by bersprockets** [comment]
@jiangxb1987 

> We shall also include #20088 in this backport PR.

I did that shortly after commenting, which allowed the tests to pass. I squashed it into the first commit, so it wasn't obvious I did it.

Should I also include #20426 in this PR, or treat that separately?

**Comment by jiangxb1987** [comment]
Both seems fine to me, it's just a minor improvement. Normally we don't backport a improvement, but since it's a simple and small change I'm confident it is safe to also include the change in a backport PR.

**Comment by bersprockets** [comment]
Hmmm... I somehow managed to break SparkR tests but fixing a comment. It seems to have auto-retried and broke the second time too.

**Comment by bersprockets** [comment]
retest this please

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        if (res != 0) return res;
+        i += 1;
+      }
+    }
+    // for architectures that support unaligned accesses, chew it up 8 bytes at a time
+    if (Platform.unaligned() || (((leftOff + i) % 8 == 0) && ((rightOff + i) % 8 == 0))) {
+      while (i <= leftLen - 8) {
+        res = (int) ((Platform.getLong(leftObj, leftOff + i) -
+                Platform.getLong(rightObj, rightOff + i)) % Integer.MAX_VALUE);
+        if (res != 0) return res;
```
Comment (mridulm) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: It is possible for two objects to be unequal and yet we consider them as equal with this code, if the long values are separated by Int.MaxValue.
Any particular reason why this code is written like this with the inaccuracy and not simply the idiomatic comparator comparison ?

File: mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala [hunk 2]
```diff
@@ -144,7 +144,7 @@ object ChiSqSelectorModel extends Loader[ChiSqSelectorModel] {
       val dataArray = Array.tabulate(model.selectedFeatures.length) { i =>
         Data(model.selectedFeatures(i))
       }
-      spark.createDataFrame(dataArray).repartition(1).write.parquet(Loader.dataPath(path))
+      spark.createDataFrame(sc.makeRDD(dataArray, 1)).write.parquet(Loader.dataPath(path))
```
Comment (dongjoon-hyun) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala]: @jiangxb1987 and @bersprockets . SPARK-22905 consists of two commits. 
- ChiSqSelector (https://github.com/apache/spark/pull/20088)
- GaussianMixtureModel (https://github.com/apache/spark/pull/20113)

If we want to include SPARK-22905 here, it had better be explicit and complete by putting `[SPARK-22905]` into the PR title and includes both patches.

Comment (bersprockets) [reply mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala]: @dongjoon-hyun Thanks, I will include the other commit and also update the title.

**Comment by squito** [comment]
overall I'm in favor of backporting this, and it looks like the only changes to the original were very small, so I'm in favor of this.

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java (hunk 1 above)
Comment (squito) [reply sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: good question, @jiangxb1987 any thoughts on this?  (it was this way in the original)

Comment (jiangxb1987) [reply sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: As far as I can recall It's actually a mistake. At first the function returned a Long value and later I changed the return value to Integer, let me fix it. Thanks for discovering this @mridulm !

**Comment by bersprockets** [comment]
@jiangxb1987 

Here are some of the differences from the original PR
- I also ported the follow up PR #20426
- I ported #20088 (for SPARK-22905) to get the tests to pass. I also ported its followup, #20113.
- sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java line 112
UnsafeExternalSorter.create does not take a Supplier for the 5th argument, so I put get() on the argument to directly pass a RecordComparator object.
- sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchange.scala: I access SQLConf differently (since SQLConf.get doesn't appear to work in 2.2).

**Comment by bersprockets** [comment]
Once this is merged, I will also back-port:

- [[SPARK-24564][TEST] Add test suite for RecordBinaryComparator](https://github.com/apache/spark/commit/5b0596648854c0c733b7c607661b78af7df18b89) 
- #22101

**Comment by bersprockets** [comment]
@jiangxb1987 gentle ping.

**Comment by bersprockets** [comment]
@gatorsmile So I should include all the related PRs merged to master as a single PR here? Just verifying.

**Comment by gatorsmile** [comment]
ok to test

**Comment by jiangxb1987** [comment]
LGTM, thanks!

**Comment by gatorsmile** [comment]
Thanks! Merged to 2.2

**Comment by bersprockets** [comment]
@gatorsmile Weird, I don't see it on branch-2.2. Is that a sync issue?

**Comment by gatorsmile** [comment]
@bersprockets Could you please close this PR?

**Pull request closed without merging**